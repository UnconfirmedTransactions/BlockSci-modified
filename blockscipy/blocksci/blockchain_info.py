import requests
import pandas as pd

class BlockchainInfoData(object):
    def __init__(self, api_key, cache_blocks=False, cache_txs=False):
        self.base_url = 'https://blockchain.info/'
        self.api_key = api_key
        self.cache_blocks = cache_blocks
        self.cache_txs = cache_txs
        self.block_cache = {}
        self.tx_cache = {}
        
    def block_data(self, block_hash):
        if block_hash in self.block_cache:
            return self.block_cache[block_hash]       
        
        url = self.base_url + 'rawblock/' + block_hash + "?api_code=" + self.api_key
        r = requests.get(url)
        r.raise_for_status()
        
        data = r.json()
        if self.cache_blocks:
            self.block_cache[block_hash] = data
        return data
    
    def tx_data(self, tx_hash):
        if tx_hash in self.tx_cache:
            return self.tx_cache[tx_hash]

        url = self.base_url + 'rawtx/' + tx_hash + "?api_code=" + self.api_key
        r = requests.get(url)
        r.raise_for_status()
        
        data = r.json()
        if self.cache_txs:
            self.tx_cache[tx_hash] = data
        return data
    
    def block_arrival_time(self, block_hash):
        data = self.block_data(block_hash)
        
        if "received_time" in data:
            return pd.to_datetime(data["received_time"], unit='s')
        else:
            return None
    
    def tx_arrival_times(self, block_hash):
        data = self.block_data(block_hash)
        arrival_times = {}
        for tx in data["tx"]:
            arrival_times[tx["hash"]] = pd.to_datetime(tx["time"], unit='s')
        return arrival_times
    
    def tx_arrival_time(self, tx_hash):
        data = self.tx_data(tx_hash)
        if "time" in data:
            return pd.to_datetime(data["time"], unit="s")
        else:
            return None
    
    def block_relayed_by(self, block_hash):
        data = self.block_data(block_hash)
        if "relayed_by" in data:
            return data["relayed_by"]
        else:
            return None