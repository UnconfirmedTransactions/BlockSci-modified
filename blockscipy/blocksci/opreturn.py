# -*- coding: utf-8 -*-
"""BlockSci Module

Label op return data as related to a specific application
"""

import binascii
import re
import Crypto.Cipher.ARC4

OP_RETURN_SERVICES = {
    "OA": "Open Assets", "id": "Blockstack",
    "S1": "Stampery", "S2": "Stampery",
    "S3": "Stampery", "S4": "Stampery", "S5": "Stampery",
    "Factom!!": "Factom", "FACTOM00": "Factom", "Fa": "Factom", "FA": "Factom",
    "ORIGMY": "OriginalMy",
    "ProveBit": "ProveBit",
    "UNicDC": "University of Nicosia",
    "DOCPROOF": "Proof of Existence",
    "omni": "Omni Layer", "ASCRIBESPOOL": "Ascribe",
    "CC": "Colu", "MG": "Monegraph",
    "SB.D": "SB.D", "BITPROOF": "Bitproof",
    "KC": "KC", "BS": "BlockSign", "OC": "UnknownOC",
    "CryptoTests-": "Crypto Copyright", "CryptoProof-": "Crypto Copyright",
    "LaPreuve": "LaPreuve",
    "RMBd": "Remembr", "RMBe": "Remembr",
    "Mined by 1hash.com": "1hash",
    "FluxST": "FluxST", "CP110400": "CP1", "KMD": "Komodo", "OKT": "OKT"
}

BYTE_PREFIXES = {
    b'g\x01\xdd6\x15&]+': "UnknownBytePrefix",
    binascii.unhexlify(b'5888'): "Blockstack",
    binascii.unhexlify(b'5808'): "Blockstack",
    binascii.unhexlify(b'455720'): "Eternity Wall",
    binascii.unhexlify(b'53504b'): "Coinspark",
    binascii.unhexlify(b'4f43'): "Openchain",
    b'STAMPD##': "stampd",
    b'CNTRPRTY': "Counterparty"
}
EXACT_STRING_MATCHES = [
    "http://www.blockcypher.com/",
    "http://www.blockcypher.com",
    "DS",
    "503: Bitcoin over capacity!",
    "XX",
    "XY",
    "XW",
    "SS"
]
EXACT_BYTE_MATCHES = [
    b'\xb3\x87\xfc~\xf6K3\xcf\x01\x82\xec\xf8\xea\xb2\x065\x8cz\xc7\xb5\xa9\xd1$\x1a\x11\xd4Sb\xda\x9af\xa2\xb9l\xb6\xb6\x9b\xd7\xa6a',
    b"\xe3\xb0\xc4B\x98\xfc\x1c\x14\x9a\xfb\xf4\xc8\x99o\xb9$'\xaeA\xe4d\x9b\x93L\xa4\x95\x99\x1bxR\xb8U",
    b'\xfc\xf4=\xa2D\x99GZ\x96\x00#1\xb8\x97\xc8\xf3#+(\xe5z\xfd\xcb?}y\x91\xbfv\xc8\xf0\xb6\x05\x81\xa2\x89\x9d\x80\x9f\xa9',
    b'*K\x94V\x827C\xfe\xc9\xc0\x7fv:\x9f\x0c\xa3@~\xba\xb2\xfa\xca\xe4\x0e\xbbm\x83\xc8\xf5k\x0b\xbf\x94\xcbn\xe7\xb0\x9a$\r'
]

# address_matches = {
#     Address(155316552, address_type.pubkeyhash):"Chainpoint", # 12m3EH1kCFWc9WvHmtLoby17xuYnSdNVYR 
#     Address(89427334, address_type.pubkeyhash):"Base64DataAddress1",
#     Address(156665979, address_type.pubkeyhash):"Base64DataAddress2",
#     Address(182075165, address_type.pubkeyhash):"Base64DataAddress3",
#     Address(156599323, address_type.pubkeyhash):"Base64DataAddress4",
#     Address(203933727, address_type.pubkeyhash):"Komodo",
#     Address(203831434, address_type.pubkey):"Komodo",
#     Address(127559021, address_type.pubkeyhash): "UnknownAddress1",
#     Address(16179390, address_type.scripthash):"SpamAddress1",
#     Address(16163151, address_type.scripthash):"SpamAddress2",
#     Address(89403362, address_type.pubkeyhash):"SpamAddress3",
#     Address(155249896, address_type.pubkeyhash):"SpamAddress4",
#     Address(77604537, address_type.pubkeyhash):"SpamAddress5",
#     Address(127504441, address_type.pubkeyhash):"SpamAddress5"
# }


def label_application(tx):
    if tx.op_return is None:
        return None
    data = tx.op_return.address.data

    if not data:
        return "Empty"

    decoded = data.decode("ascii", "replace")
    strings = re.findall("[a-zA-Z0-9.://! ]+", decoded)

    if decoded in EXACT_STRING_MATCHES:
        return decoded
    elif data in EXACT_BYTE_MATCHES:
        return data
    elif strings and strings[0].startswith(tuple(OP_RETURN_SERVICES.keys())):
        for prefix in OP_RETURN_SERVICES:
            if strings[0].startswith(prefix):
                return OP_RETURN_SERVICES[prefix]
    elif data.startswith(tuple(BYTE_PREFIXES.keys())):
        for prefix in BYTE_PREFIXES:
            if data.startswith(prefix):
                return BYTE_PREFIXES[prefix]
    # elif len(set(txout.address for txout in tx.outs).intersection(set(address_matches.keys()))) > 0:
    #     for txout in tx.outs:
    #         if txout.address in address_matches:
    #             return address_matches[txout.address]
    elif tx.ins:
        first_vin_txid = binascii.unhexlify(str(tx.ins[0].spent_tx.hash))
        decoded = Crypto.Cipher.ARC4.new(first_vin_txid).decrypt(data)
        if decoded.startswith(b'CNTRPRTY'):
            return "Counterparty"
    return "Unknown"
