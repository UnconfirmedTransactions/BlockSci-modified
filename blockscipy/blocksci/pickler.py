from ._blocksci import *
import pickle
import sqlite3
from collections import namedtuple

# Simple class representing a record in our database.
MemoRecord = namedtuple("MemoRecord", "key, task")

class Pickler(pickle.Pickler):
    """Custom Pickler for BlockSci objects"""

    def persistent_id(self, obj):
        # Instead of pickling MemoRecord as a regular class instance, we emit a
        # persistent ID.
        if isinstance(obj, Block):
            # Here, our persistent ID is simply a tuple, containing a tag and a
            # key, which refers to a specific record in the database.
            return ("Block", obj.height)
        elif isinstance(obj, Tx):
            return ("Tx", obj.index)
        elif isinstance(obj, Output):
            return ("Output", (obj.tx_index, obj.index))
        elif isinstance(obj, Input):
            return ("Input", (obj.tx_index, obj.index))
        elif isinstance(obj, Address):
            return ("Address", (obj.address_num, obj.type))
        elif isinstance(obj, EquivAddress):
            sample_addr = next(iter(obj))
            return ("EquivAddress", (sample_addr.address_num, sample_addr.type, obj.is_script_equiv))
        else:
            # If obj does not have a persistent ID, return None. This means obj
            # needs to be pickled as usual.
            return None


class Unpickler(pickle.Unpickler):
    """Custom Unpickler for BlockSci objects"""

    def __init__(self, file, chain):
        super().__init__(file)
        self.chain = chain

    def persistent_load(self, pid):
        # This method is invoked whenever a persistent ID is encountered.
        # Here, pid is the tuple returned by DBPickler.
        type_tag, key_id = pid
        if type_tag == "Block":
            return self.chain[key_id]
        elif type_tag == "Tx":
            return self.chain.tx_with_index(key_id)
        elif type_tag == "Output":
            return self.chain.tx_with_index(key_id[0]).outputs[key_id[1]]
        elif type_tag == "Input":
            return self.chain.tx_with_index(key_id[0]).inputs[key_id[1]]
        elif type_tag == "Address":
            return self.chain.address_from_index(key_id[0], key_id[1])
        elif type_tag == "EquivAddress":
            addr = self.chain.address_from_index(key_id[0], key_id[1])
            return addr.equiv(key_id[2])
        else:
            # Always raises an error if you cannot return the correct object.
            # Otherwise, the unpickler will think None is the object referenced
            # by the persistent ID.
            raise pickle.UnpicklingError("unsupported persistent object")
