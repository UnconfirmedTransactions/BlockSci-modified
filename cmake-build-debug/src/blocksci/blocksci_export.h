
#ifndef BLOCKSCI_EXPORT_H
#define BLOCKSCI_EXPORT_H

#ifdef BLOCKSCI_STATIC_DEFINE
#  define BLOCKSCI_EXPORT
#  define BLOCKSCI_NO_EXPORT
#else
#  ifndef BLOCKSCI_EXPORT
#    ifdef blocksci_EXPORTS
        /* We are building this library */
#      define BLOCKSCI_EXPORT __attribute__((visibility("default")))
#    else
        /* We are using this library */
#      define BLOCKSCI_EXPORT __attribute__((visibility("default")))
#    endif
#  endif

#  ifndef BLOCKSCI_NO_EXPORT
#    define BLOCKSCI_NO_EXPORT __attribute__((visibility("hidden")))
#  endif
#endif

#ifndef BLOCKSCI_DEPRECATED
#  define BLOCKSCI_DEPRECATED __attribute__ ((__deprecated__))
#endif

#ifndef BLOCKSCI_DEPRECATED_EXPORT
#  define BLOCKSCI_DEPRECATED_EXPORT BLOCKSCI_EXPORT BLOCKSCI_DEPRECATED
#endif

#ifndef BLOCKSCI_DEPRECATED_NO_EXPORT
#  define BLOCKSCI_DEPRECATED_NO_EXPORT BLOCKSCI_NO_EXPORT BLOCKSCI_DEPRECATED
#endif

#if 0 /* DEFINE_NO_DEPRECATED */
#  ifndef BLOCKSCI_NO_DEPRECATED
#    define BLOCKSCI_NO_DEPRECATED
#  endif
#endif

#endif /* BLOCKSCI_EXPORT_H */
