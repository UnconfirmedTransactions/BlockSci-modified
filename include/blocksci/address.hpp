//
//  address.hpp
//  blocksci
//
//  Created by Harry Kalodner on 12/12/17.
//
//

#ifndef blocksci_address_group_header
#define blocksci_address_group_header

#include <blocksci/address/address.hpp>
#include <blocksci/address/equiv_address.hpp>

#endif /* blocksci_address_group_header */
