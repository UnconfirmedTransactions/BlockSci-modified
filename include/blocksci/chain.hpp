//
//  chain.hpp
//  blocksci
//
//  Created by Harry Kalodner on 9/4/17.
//
//

#ifndef chain_h
#define chain_h

#include <blocksci/chain/algorithms.hpp>
#include <blocksci/chain/block.hpp>
#include <blocksci/chain/blockchain.hpp>
#include <blocksci/chain/input_pointer.hpp>
#include <blocksci/chain/input.hpp>
#include <blocksci/chain/output_pointer.hpp>
#include <blocksci/chain/output.hpp>
#include <blocksci/chain/transaction.hpp>
#include <blocksci/chain/transaction_range.hpp>

#endif /* chain_h */
