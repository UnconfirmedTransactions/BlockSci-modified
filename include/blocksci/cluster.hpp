//
//  cluster.h
//  blocksci
//
//  Created by Harry Kalodner on 4/4/18.
//

#ifndef cluster_group_header_hpp
#define cluster_group_header_hpp

#include <blocksci/cluster/cluster.hpp>
#include <blocksci/cluster/cluster_manager.hpp>

#endif /* cluster_group_header_hpp */
