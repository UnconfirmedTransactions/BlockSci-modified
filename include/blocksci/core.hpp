//
//  core.hpp
//  blocksci
//
//  Created by Harry Kalodner on 4/22/18.
//
//

#ifndef core_group_header_h
#define core_group_header_h

#include <blocksci/core/bitcoin_uint256.hpp>
#include <blocksci/core/script_data.hpp>
#include <blocksci/core/address_types.hpp>
#include <blocksci/core/inout.hpp>
#include <blocksci/core/raw_transaction.hpp>
#include <blocksci/core/raw_block.hpp>


#endif /* core_group_header_h */
