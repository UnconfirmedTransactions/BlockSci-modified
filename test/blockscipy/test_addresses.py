import blocksci
import pytest
from util import Coin, sorted_tx_list


def address_types(chain_name):
    if chain_name == "btc":
        return ["p2pkh", "p2sh", "p2wpkh", "p2wsh"]
    else:
        return ["p2pkh", "p2sh"]


def addresses(chain, json_data, chain_name):
    for addr_type in address_types(chain_name):
        for i in range(3):
            addr = chain.address_from_string(
                json_data["address-{}-spend-{}".format(addr_type, i)]
            )
            yield addr, addr_type

    raw_multisig_tx = chain.tx_with_hash(json_data["raw-multisig-tx"])
    yield raw_multisig_tx.outputs[0].address, "raw-multisig"

    p2sh_multisig_tx = chain.tx_with_hash(json_data["p2sh-multisig-tx"])
    yield p2sh_multisig_tx.outputs[0].address, "p2sh-multisig"


def address_received_test(addr, address_type, balance, ntxes):
    assert address_type == addr.type
    assert Coin(balance) == addr.balance()
    assert ntxes == len(addr.txes.to_list())


def test_p2pkh_address(chain, json_data):
    addr = chain.address_from_string(json_data["address-p2pkh-spend-0"])
    address_received_test(addr, blocksci.address_type.pubkeyhash, 1, 1)

    addr = chain.address_from_string(json_data["address-p2pkh-spend-1"])
    address_received_test(addr, blocksci.address_type.pubkeyhash, 0, 2)

    addr = chain.address_from_string(json_data["address-p2pkh-spend-2"])
    address_received_test(addr, blocksci.address_type.pubkeyhash, 0, 4)


def test_p2sh_address(chain, json_data):
    addr = chain.address_from_string(json_data["address-p2sh-spend-0"])
    address_received_test(addr, blocksci.address_type.scripthash, 1, 1)

    addr = chain.address_from_string(json_data["address-p2sh-spend-1"])
    address_received_test(addr, blocksci.address_type.scripthash, 0, 2)

    addr = chain.address_from_string(json_data["address-p2sh-spend-2"])
    address_received_test(addr, blocksci.address_type.scripthash, 0, 4)


@pytest.mark.btc
@pytest.mark.ltc
def test_p2wpkh_address(chain, json_data):
    addr = chain.address_from_string(json_data["address-p2wpkh-spend-0"])
    address_received_test(addr, blocksci.address_type.witness_pubkeyhash, 1, 1)

    addr = chain.address_from_string(json_data["address-p2wpkh-spend-1"])
    address_received_test(addr, blocksci.address_type.witness_pubkeyhash, 0, 2)

    addr = chain.address_from_string(json_data["address-p2wpkh-spend-2"])
    address_received_test(addr, blocksci.address_type.witness_pubkeyhash, 0, 4)


@pytest.mark.btc
@pytest.mark.ltc
def test_p2wsh_address(chain, json_data):
    addr = chain.address_from_string(json_data["address-p2wsh-spend-0"])
    address_received_test(addr, blocksci.address_type.witness_scripthash, 1, 1)

    addr = chain.address_from_string(json_data["address-p2wsh-spend-1"])
    address_received_test(addr, blocksci.address_type.witness_scripthash, 0, 2)

    addr = chain.address_from_string(json_data["address-p2wsh-spend-2"])
    address_received_test(addr, blocksci.address_type.witness_scripthash, 0, 4)


def test_address_regression(chain, json_data, regtest, chain_name):
    for addr, addr_type in addresses(chain, json_data, chain_name):
        if "multisig" not in addr_type:
            print(addr.address_string, file=regtest)
        print(addr.balance(), file=regtest)
        print(addr.equiv(), file=regtest)
        print(addr.has_been_spent, file=regtest)
        print(addr.inputs.to_list(), file=regtest)
        print(sorted_tx_list(addr.input_txes), file=regtest)
        print(addr.input_txes_count(), file=regtest)
        print(sorted_tx_list(addr.output_txes), file=regtest)
        print(addr.output_txes_count(), file=regtest)
        print(addr.outputs.to_list(), file=regtest)
        if addr_type == "p2pkh" or addr_type == "p2wpkh":
            print(addr.pubkey, file=regtest)
            print(addr.pubkeyhash, file=regtest)
        print(addr.raw_type, file=regtest)
        print(addr.revealed_tx, file=regtest)
        print(addr.type, file=regtest)
        print(addr.full_type, file=regtest)
        if addr.type == blocksci.address_type.multisig:
            print(addr.required, file=regtest)
            print(addr.total, file=regtest)
        if "p2sh" in addr_type or addr_type == "p2wsh":
            print(addr.wrapped_address, file=regtest)
