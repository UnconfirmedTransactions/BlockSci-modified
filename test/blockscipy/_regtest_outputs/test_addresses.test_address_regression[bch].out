myowM8GghMuCX4tn58Nmh9eubwMjuwKiCJ
100000000
EquivAddress(PubkeyHashAddress(myowM8GghMuCX4tn58Nmh9eubwMjuwKiCJ))
False
[]
[]
0
[Tx(len(txins)=1, len(txouts)=2, size_bytes=226, block_height=111, tx_index=112)]
1
[TxOut(spending_tx_index=0, address=PubkeyHashAddress(myowM8GghMuCX4tn58Nmh9eubwMjuwKiCJ), value=100000000)]
None
7fd8abeed9bbe533e8bac0d38732dd9c7f8ba8c8
2
None
Pay to pubkey hash
pubkeyhash
mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T
0
EquivAddress(PubkeyHashAddress(mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T))
True
[TxIn(spent_tx_index=115, address=PubkeyHashAddress(mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T), value=100000000)]
[Tx(len(txins)=1, len(txouts)=1, size_bytes=191, block_height=112, tx_index=114)]
1
[Tx(len(txins)=1, len(txouts)=2, size_bytes=225, block_height=112, tx_index=115)]
1
[TxOut(spending_tx_index=114, address=PubkeyHashAddress(mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T), value=100000000)]
b'\x02\x9e\xc4{e\xef\xbd\xd2N\xbcp\x10\xfbg\x1cz\xab\x92\x8e\x91"4?9\rpq\xda\x15\xe5l\xf5!'
dd9f85e63f3e6f87bb9ee130bb2c8a62dea33c77
2
Tx(len(txins)=1, len(txouts)=1, size_bytes=191, block_height=112, tx_index=114)
Pay to pubkey hash
pubkeyhash
mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb
0
EquivAddress(PubkeyHashAddress(mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb))
True
[TxIn(spent_tx_index=117, address=PubkeyHashAddress(mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb), value=200000000), TxIn(spent_tx_index=118, address=PubkeyHashAddress(mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb), value=100000000)]
[Tx(len(txins)=1, len(txouts)=1, size_bytes=191, block_height=113, tx_index=119), Tx(len(txins)=1, len(txouts)=1, size_bytes=191, block_height=113, tx_index=120)]
2
[Tx(len(txins)=1, len(txouts)=2, size_bytes=226, block_height=113, tx_index=117), Tx(len(txins)=1, len(txouts)=2, size_bytes=226, block_height=113, tx_index=118)]
2
[TxOut(spending_tx_index=120, address=PubkeyHashAddress(mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb), value=200000000), TxOut(spending_tx_index=119, address=PubkeyHashAddress(mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb), value=100000000)]
b'\x03\x08\xde\xe8\xbb\xccv\xa7\xad\xdfm\xd7\xa2\xdam\xbf\xa8\xa0N(\x03\xd1\r\x98\x8c)\x8d\x12\x9f\xd3l\xae\xea'
9a337fd549963efeb8780c93e234bbf754e15d32
2
Tx(len(txins)=1, len(txouts)=1, size_bytes=191, block_height=113, tx_index=119)
Pay to pubkey hash
pubkeyhash
2Mshjw3i5VcfqCm6FaJLxLSdSQhShKJYk6d
100000000
EquivAddress(ScriptHashAddress(2Mshjw3i5VcfqCm6FaJLxLSdSQhShKJYk6d))
False
[]
[]
0
[Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=114, tx_index=122)]
1
[TxOut(spending_tx_index=0, address=ScriptHashAddress(2Mshjw3i5VcfqCm6FaJLxLSdSQhShKJYk6d), value=100000000)]
4
None
Pay to script hash
scripthash
None
2Mzqim7c3TBkpRyqXmB23DXXcm1nHN6kFsf
0
EquivAddress(PubkeyAddress(mr6VKoQ9zMau7CSNaf1QBzy6Pw3DmMPrWm), ScriptHashAddress(2Mzqim7c3TBkpRyqXmB23DXXcm1nHN6kFsf))
True
[TxIn(spent_tx_index=125, address=ScriptHashAddress(2Mzqim7c3TBkpRyqXmB23DXXcm1nHN6kFsf), value=100000000)]
[Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=115, tx_index=124)]
1
[Tx(len(txins)=1, len(txouts)=2, size_bytes=224, block_height=115, tx_index=125)]
1
[TxOut(spending_tx_index=124, address=ScriptHashAddress(2Mzqim7c3TBkpRyqXmB23DXXcm1nHN6kFsf), value=100000000)]
4
Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=115, tx_index=124)
Pay to script hash
scripthash/pubkey
PubkeyAddress(mr6VKoQ9zMau7CSNaf1QBzy6Pw3DmMPrWm)
2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW
0
EquivAddress(PubkeyAddress(mjAzMtZc7UFuvuRqJyhrrVXHQKoY5UhG6R), ScriptHashAddress(2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW))
True
[TxIn(spent_tx_index=128, address=ScriptHashAddress(2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW), value=100000000), TxIn(spent_tx_index=129, address=ScriptHashAddress(2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW), value=200000000)]
[Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=116, tx_index=127), Tx(len(txins)=1, len(txouts)=1, size_bytes=194, block_height=116, tx_index=130)]
2
[Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=116, tx_index=128), Tx(len(txins)=1, len(txouts)=2, size_bytes=224, block_height=116, tx_index=129)]
2
[TxOut(spending_tx_index=130, address=ScriptHashAddress(2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW), value=100000000), TxOut(spending_tx_index=127, address=ScriptHashAddress(2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW), value=200000000)]
4
Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=116, tx_index=127)
Pay to script hash
scripthash/pubkey
PubkeyAddress(mjAzMtZc7UFuvuRqJyhrrVXHQKoY5UhG6R)
0
EquivAddress(MultisigAddress(2 of 3))
True
[TxIn(spent_tx_index=304, address=MultisigAddress(2 of 3), value=9990000)]
[Tx(len(txins)=1, len(txouts)=1, size_bytes=231, block_height=183, tx_index=303)]
1
[Tx(len(txins)=1, len(txouts)=1, size_bytes=271, block_height=183, tx_index=304)]
1
[TxOut(spending_tx_index=303, address=MultisigAddress(2 of 3), value=9990000)]
5
Tx(len(txins)=1, len(txouts)=1, size_bytes=231, block_height=183, tx_index=303)
Multisig
multisig2of3
2
3
0
EquivAddress(ScriptHashAddress(2N2PGeTV9hPu4nPtJHgNmV9shPRxKJy9Xqj), MultisigAddress(2 of 3))
True
[TxIn(spent_tx_index=309, address=ScriptHashAddress(2N2PGeTV9hPu4nPtJHgNmV9shPRxKJy9Xqj), value=9990000)]
[Tx(len(txins)=1, len(txouts)=1, size_bytes=341, block_height=184, tx_index=307)]
1
[Tx(len(txins)=1, len(txouts)=1, size_bytes=189, block_height=184, tx_index=309)]
1
[TxOut(spending_tx_index=307, address=ScriptHashAddress(2N2PGeTV9hPu4nPtJHgNmV9shPRxKJy9Xqj), value=9990000)]
4
Tx(len(txins)=1, len(txouts)=1, size_bytes=341, block_height=184, tx_index=307)
Pay to script hash
scripthash/multisig2of3
MultisigAddress(2 of 3 multisig with addresses mrfx8rZk4kfkce2Dvxy1m5rcGVRmyZGn1U, n2vmQHZ1wK4meS4wvLpMNhdgwaderv8Uij, mr9KKBWmDqCWb7hpELAGnUyvYPbbMGpiSM)
